# This patch file is from Qualcomm Technologies, Inc. and is provided under the following license:
#
# Copyright (c) 2025 Qualcomm Technologies, Inc. All rights reserved.
#
# SPDX-License-Identifier: BSD-3-Clause-Clear

From e848be0d6c888a3c69aaa5baa0ef6e9fb53d02ca Mon Sep 17 00:00:00 2001
From: Ulrich Meis <umeis@qti.qualcomm.com>
Date: Tue, 30 Apr 2019 13:24:42 +0200
Subject: [PATCH] BootLinux: use fastboot header info if kernel_addr >=
 0xa0000000.

This enables loading of images at locations specified in the fastboot header.
It is required to load the COQOS hypervisor at a different address than the
Linux.

Change-Id: I03eb0ee91630869fbc18fda7654cec4a7425da35
Signed-off-by: Ankurkumar Satasiya <asatasiy@qti.qualcomm.com>
---
 QcomModulePkg/Library/BootLib/BootLinux.c | 23 ++++++++++++++++++++---
 1 file changed, 20 insertions(+), 3 deletions(-)

diff --git a/QcomModulePkg/Library/BootLib/BootLinux.c b/QcomModulePkg/Library/BootLib/BootLinux.c
index fc4a7ec..4c60f66 100644
--- a/QcomModulePkg/Library/BootLib/BootLinux.c
+++ b/QcomModulePkg/Library/BootLib/BootLinux.c
@@ -1116,10 +1116,27 @@ skip_FfbmStr:
     return EFI_BAD_BUFFER_SIZE;
   }
 
-  Status = UpdateBootParams (&BootParamlistPtr, FlashlessBoot);
-  if (Status != EFI_SUCCESS) {
-    return Status;
+  /* Use information in fastboot header if kernel address >= 0xa0000000 */
+  if (((boot_img_hdr *)(BootParamlistPtr.ImageBuffer))->kernel_addr < 0xa0000000) {
+    Status = UpdateBootParams (&BootParamlistPtr, FlashlessBoot);
+    if (Status != EFI_SUCCESS) {
+      return Status;
+    }
+  } else {
+    UINT64 KernelSizeReserved;
+
+    KernelSizeReserved = PcdGet32 (RamdiskEndAddress);
+
+    BootParamlistPtr.KernelLoadAddr =
+                 ((boot_img_hdr *)(BootParamlistPtr.ImageBuffer))->kernel_addr;
+    BootParamlistPtr.RamdiskLoadAddr =
+                 ((boot_img_hdr *)(BootParamlistPtr.ImageBuffer))->ramdisk_addr;
+    BootParamlistPtr.DeviceTreeLoadAddr =
+                 ((boot_img_hdr *)(BootParamlistPtr.ImageBuffer))->tags_addr;
+    BootParamlistPtr.KernelEndAddr = BootParamlistPtr.BaseMemory +
+                                       KernelSizeReserved;
   }
+
   SetandGetLoadAddr (&BootParamlistPtr, LOAD_ADDR_NONE);
   Status = GZipPkgCheck (&BootParamlistPtr);
   if (Status != EFI_SUCCESS) {
-- 
2.7.4

